### PR TITLE
docs: Adds developer docs to repo: Fixes #1648 Fixes #1649 Fixes #165…

### LINKED ISSUE #165: Possible bug: Google_Service_AndroidPublisher can't be used with Service Account Settings of OAuths 2
Hi
I am trying to verity a purchase of inapppurchase products via my backend server but i need to setup on OAuths 2 as Service Account to use Google_Service_AndroidPublisher($client); 
without human interactions, it seems it is going to work just with Web application Accounts which needs human interaction.

please clarify it here, if this is a bug or you have it in the road map.
Thanks

http://stackoverflow.com/questions/22513922/403-forbidden-response-from-google-android-in-app-purchases-api-with-service-a

http://milancermak.wordpress.com/2012/08/24/server-side-verification-of-google-play-subsc/

### LINKED ISSUE #1648: DevSite Import: Calling an API
Notice: This DevSite page will soon be deleted per [b/133171590](http://b/133171590). Please add this sample to GitHub if it is relevant still:

https://developers.google.com/api-client-library/php/auth/web-app#example

---

# OAuth 2.0 Complete Example

`index.php`
```php
<?php
require_once __DIR__.'/vendor/autoload.php';

session_start();

$client = new Google_Client();
$client->setAuthConfig('client_secrets.json');
$client->addScope(Google_Service_Drive::DRIVE_METADATA_READONLY);

if (isset($_SESSION['access_token']) && $_SESSION['access_token']) {
  $client->setAccessToken($_SESSION['access_token']);
  $drive = new Google_Service_Drive($client);
  $files = $drive->files->listFiles(array())->getItems();
  echo json_encode($files);
} else {
  $redirect_uri = 'http://' . $_SERVER['HTTP_HOST'] . '/oauth2callback.php';
  header('Location: ' . filter_var($redirect_uri, FILTER_SANITIZE_URL));
}
```

`oauth2callback.php`
```php
<?php
require_once __DIR__.'/vendor/autoload.php';

session_start();

$client = new Google_Client();
$client->setAuthConfigFile('client_secrets.json');
$client->setRedirectUri('http://' . $_SERVER['HTTP_HOST'] . '/oauth2callback.php');
$client->addScope(Google_Service_Drive::DRIVE_METADATA_READONLY);

if (! isset($_GET['code'])) {
  $auth_url = $client->createAuthUrl();
  header('Location: ' . filter_var($auth_url, FILTER_SANITIZE_URL));
} else {
  $client->authenticate($_GET['code']);
  $_SESSION['access_token'] = $client->getAccessToken();
  $redirect_uri = 'http://' . $_SERVER['HTTP_HOST'] . '/';
  header('Location: ' . filter_var($redirect_uri, FILTER_SANITIZE_URL));
}
```

### LINKED ISSUE #1649: DevSite Import: Authentication
Notice: This DevSite page will soon be deleted per [b/133171590](http://b/133171590). Please add this sample to GitHub if it is relevant still:

https://developers.google.com/api-client-library/php/guide/aaa_idtoken

---

# ID Token Authentication

ID tokens allow authenticating a user securely without requiring a network call (in many cases), and without granting the server access to request user information from the Google APIs.

For a complete example, see the idtoken.php sample in the examples/ directory of the client library.
This is accomplished because each ID token is a cryptographically signed, base64 encoded JSON structure. The token payload includes the Google user ID, the client ID of the application the user signed in to, and the issuer (in this case, Google). It also contains a cryptographic signature which can be verified with the public Google certificates to ensure that the token was created by Google. If the user has granted permission to view their email address to the application, the ID token will additionally include their email address.

The token can be easily and securely verified with the PHP client library

```php
function getUserFromToken($token) {
  $ticket = $client->verifyIdToken($token);
  if ($ticket) {
    $data = $ticket->getAttributes();
    return $data['payload']['sub']; // user ID
  }
  return false
}
```

The library will automatically download and cache the certificate required for verification, and refresh it if it has expired.

### PR DESCRIPTION
**Pull request by grant**
Adds a documentation folder that is imported from developers.google.com
Content hasn't been changed besides converting to markdown.

---

R: @bshaffer

### CODE REVIEW

File: docs/oauth-server.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+Use the authorized `Google_Client` object to call Google APIs by completing the following steps:
+
+1.  Build a service object for the API that you want to call, providing the authorized `Google_Client` object. For example, to call the Cloud SQL Administration API:
+    
+    ```php
+    $sqladmin = new Google_Service_SQLAdmin($client);
+    ```
+    
+2.  Make requests to the API service using the [interface provided by the service object](https://devsite.googleplex.com/api-client-library/php/start/get_started#build). For example, to list the instances of Cloud SQL databases in the examinable-example-123 project:
```
Comment (dwsupplee) [inline docs/oauth-server.md]: This should be going [here](https://github.com/googleapis/google-api-php-client/pull/1668/files#diff-e2348c79e582266d6ee3351360bccf3dR62)

File: docs/api-keys.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# API Keys
+
+When calling APIs that do not access private user data, you can use simple API keys. These keys are used to authenticate your application for accounting purposes. The Google Developers Console documentation also describes [API keys](https://developers.google.com/console/help/using-keys).
+
+> Note: If you do need to access private user data, you must use OAuth 2.0. See [Using OAuth 2.0 for Web Server Applications and Using OAuth 2.0 for Server to Server Applications](docs/oauth.md) for more information.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/api-keys.md]: Done.

File: docs/start.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+It is important to understand the basics of how API authentication and authorization are handled. All API calls must use either simple or authorized access (defined below). Many API methods require authorized access, but some can use either. Some API methods that can use either behave differently, depending on whether you use simple or authorized access. See the API's method documentation to determine the appropriate access type.
+
+### 1. Simple API access (API keys)
+
+These API calls do not access any private user data. Your application must authenticate itself as an application belonging to your Google Cloud project. This is needed to measure project usage for accounting purposes.
+
+#### Important concepts
+
+*   **API key**: To authenticate your application, use an [API key](/console/help/#WhatIsKey) for your Google Cloud Console project. Every simple access call your application makes must include this key.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/start.md]: Done.

File: docs/start.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    
+> **Warning**: Keep your API key private. If someone obtains your key, they could use it to consume your quota or incur charges against your Google Cloud project.
+    
+
+### 2. Authorized API access (OAuth 2.0)
+
+These API calls access private user data. Before you can call them, the user that has access to the private data must grant your application access. Therefore, your application must be authenticated, the user must grant access for your application, and the user must be authenticated in order to grant that access. All of this is accomplished with [OAuth 2.0](/accounts/docs/OAuth2) and libraries written for it.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/start.md]: Done.

File: docs/oauth-web.md [hunk 5]
```diff
@@ -0,0 +1,424 @@
+# Using OAuth 2.0 for Web Server Applications
+
+This document explains how web server applications use the Google API Client Library for PHP to implement OAuth 2.0 authorization to access Google APIs. OAuth 2.0 allows users to share specific data with an application while keeping their usernames, passwords, and other information private. For example, an application can use OAuth 2.0 to obtain permission from users to store files in their Google Drives.
+
+This OAuth 2.0 flow is specifically for user authorization. It is designed for applications that can store confidential information and maintain state. A properly authorized web server application can access an API while the user interacts with the application or after the user has left the application.
+
+Web server applications frequently also use [service accounts](https://developers.google.com/api-client-library/php/auth/service-accounts) to authorize API requests, particularly when calling Cloud APIs to access project-based data rather than user-specific data. Web server applications can use service accounts in conjunction with user authorization.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Done.

File: docs/oauth-web.md [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+Any application that uses OAuth 2.0 to access Google APIs must have authorization credentials that identify the application to Google's OAuth 2.0 server. The following steps explain how to create credentials for your project. Your applications can then use the credentials to access APIs that you have enabled for that project.
+
+1.  Open the [Credentials page](https://console.developers.google.com/apis/credentials) in the API Console.
+2.  Click **Create credentials > OAuth client ID**.
+3.  Complete the form. Set the application type to `Web application`. Applications that use languages and frameworks like PHP, Java, Python, Ruby, and .NET must specify authorized **redirect URIs**. The redirect URIs are the endpoints to which the OAuth 2.0 server can send responses.  
+      
+    For testing, you can specify URIs that refer to the local machine, such as `http://localhost:8080`. With that in mind, please note that all of the examples in this document use `http://localhost:8080` as the redirect URI.  
+      
+    We recommend that you [design your app's auth endpoints](#protectauthcode) so that your application does not expose authorization codes to other resources on the page.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Removed the unnecessary link.

File: docs/oauth-web.md [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+
+The following rules apply to an access token obtained from an incremental authorization:
+
+*   The token can be used to access resources corresponding to any of the scopes rolled into the new, combined authorization.
+*   When you use the refresh token for the combined authorization to obtain an access token, the access token represents the combined authorization and can be used for any of its scopes.
+*   The combined authorization includes all scopes that the user granted to the API project even if the grants were requested from different clients. For example, if a user granted access to one scope using an application's desktop client and then granted another scope to the same application via a mobile client, the combined authorization would include both scopes.
+*   If you revoke a token that represents a combined authorization, access to all of that authorization's scopes on behalf of the associated user are revoked simultaneously.
+
+The example for [setting authorization parameters](#creatingclient) demonstrates how to ensure authorization requests follow this best practice. The code snippet below also shows the code that you need to add to use incremental authorization.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Fixed.

File: docs/oauth-web.md [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+4.  Your application finds out what the user decided.
+5.  If the user granted the requested permissions, your application retrieves tokens needed to make API requests on the user's behalf.
+
+### Step 1: Set authorization parameters
+
+Your first step is to create the authorization request. That request sets parameters that identify your application and define the permissions that the user will be asked to grant to your application.
+
+The code snippet below creates a `Google_Client()` object, which defines the parameters in the authorization request.
+
+That object uses information from your **client_secret.json** file to identify your application. (See [creating authorization credentials](#creatingcred) for more about that file.) The object also identifies the scopes that your application is requesting permission to access and the URL to your application's auth endpoint, which will handle the response from Google's OAuth 2.0 server. Finally, the code sets the optional access_type and include_granted_scopes parameters.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Removed.

File: docs/oauth-web.md [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+  
+Scopes enable your application to only request access to the resources that it needs while also enabling users to control the amount of access that they grant to your application. Thus, there is an inverse relationship between the number of scopes requested and the likelihood of obtaining user consent. To set this value in PHP, call the `addScope` function:
+
+```php
+$client->addScope(Google_Service_Drive::DRIVE_METADATA_READONLY);
+```
+
+The [OAuth 2.0 API Scopes](https://developers.google.com/identity/protocols/googlescopes) document provides a full list of scopes that you might use to access Google APIs.  
+  
+We recommend that your application request access to authorization scopes in context whenever possible. By requesting access to user data in context, via [incremental authorization](#incrementalAuth), you help users to more easily understand why your application needs the access it is requesting.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Done.

File: docs/oauth-web.md [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+  
+To set this value in PHP, call the `setState` function:
+
+```php
+$client->setState($sample_passthrough_value);
+```
+
+##### `include_granted_scopes`
+
+**Optional**. Enables applications to use incremental authorization to request access to additional scopes in context. If you set this parameter's value to `true` and the authorization request is granted, then the new access token will also cover any scopes to which the user previously granted the application access. See the [incremental authorization](#incrementalAuth) section for examples.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Done.

File: docs/oauth-web.md [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+
+Prompt the user for consent.
+
+`select_account`
+
+Prompt the user to select an account.
+
+### Step 2: Redirect to Google's OAuth 2.0 server
+
+Redirect the user to Google's OAuth 2.0 server to initiate the authentication and authorization process. Typically, this occurs when your application first needs to access the user's data. In the case of [incremental authorization](#incrementalAuth), this step also occurs when your application first needs to access additional resources that it does not yet have permission to access.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Done.

File: docs/oauth-web.md [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+    $client->setAccessToken($access_token);
+    ```
+    
+2.  Build a service object for the API that you want to call. You build a a service object by providing an authorized `Google_Client` object to the constructor for the API you want to call. For example, to call the Drive API:
+    
+    ```php
+    $drive = new Google_Service_Drive($client);
+    ```
+    
+3.  Make requests to the API service using the [interface provided by the service object](https://developers.google.com/api-client-library/php/start/get_started#build). For example, to list the files in the authenticated user's Google Drive:
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Done.

File: docs/oauth-web.md [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+
+```php
+$client->setIncludeGrantedScopes(true);
+```
+
+## Refreshing an access token (offline access)
+
+Access tokens periodically expire. You can refresh an access token without prompting the user for permission (including when the user is not present) if you requested offline access to the scopes associated with the token.
+
+If you use a Google API Client Library, the [client object](#creatingclient) refreshes the access token as needed as long as you configure that object for offline access.
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Done.

File: docs/start.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+*   **Refresh and access tokens**: When a user grants your application access, the OAuth 2.0 authorization server provides your application with refresh and access tokens. These tokens are only valid for the scope requested. Your application uses access tokens to authorize API calls. Access tokens expire, but refresh tokens do not. Your application can use a refresh token to acquire a new access token.
+    
+    > **Warning**: Keep refresh and access tokens private. If someone obtains your tokens, they could use them to access private user data.
+    
+*   **Client ID and client secret**: These strings uniquely identify your application and are used to acquire tokens. They are created for your Google Cloud project on the [API Access pane](https://code.google.com/apis/console#:access) of the Google Cloud. There are three types of client IDs, so be sure to get the correct type for your application:
+    
+    *   Web application client IDs
+    *   Installed application client IDs
+    *   [Service Account](/accounts/docs/OAuth2ServiceAccount) client IDs
```
[the remark this answers is not in the document]
Comment (grant) [reply docs/start.md]: Done.

**Comment by grant** [comment]
Thanks for reviewing.

> What's the status of the rollback of the doc site changes? It's unfortunate all the links out on the web no longer take people to the specific sections of these documents they are looking for.

After talking to Brent, we decided that it's better if we keep the API client library docs in this repo rather than at developers.google.com/api-client-library/php since we are able to track and fix changes here more easily.

As for the redirects, we can apply fine-grain redirects after this PR is in.

**Comment by grant** [comment]
> I'd suggest linking this to the relevant DevSite docs instead of adding this here, where it would quickly go out of date.

The point of this PR is to keep the docs close to the source code. The previous docs were last updated in ~2015. Since the libraries are updated more frequently, it's better to keep the docs here where they are visible by the team that works on the client.

File: docs/oauth-web.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (dwsupplee) [reply docs/oauth-web.md]: Looks like this was missed

[the remark this answers is not in the document]
Comment (grant) [reply docs/oauth-web.md]: Hmm, well I fixed it for sure now.

**Comment by dwsupplee** [comment]
Thanks for the quick turnaround on the review @grant

**Comment by grant** [comment]
> Thanks for the quick turnaround on the review @grant

Anytime.

**Pull request merged**